src/include/units/physical/international/length.h

```diff
@@ -0,0 +1,103 @@
+
+// The MIT License (MIT)
+//
+// Copyright (c) 2018 Mateusz Pusz
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#pragma once
+
+#include <units/physical/si/length.h>
+
+/*
+ international units of length are defined in terms of the si metre
+*/
+
+namespace units::international {
+
+//international yard
+//https://en.wikipedia.org/wiki/International_yard_and_pound
+struct yard : named_scaled_unit<yard, "yd", no_prefix, ratio<9'144, 1'000,-1>, si::metre> {};
+
+
+//international foot
+//https://en.wikipedia.org/wiki/Foot_(unit)#International_foot
+struct foot : named_scaled_unit<foot, "ft", no_prefix, ratio<3'048,1'000,-1>, si::metre> {};
+
+//https://en.wikipedia.org/wiki/Fathom#International_fathom
+struct fathom : named_scaled_unit<fathom, "fathom",no_prefix,ratio<1'828,1'000>,si::metre > {};
+
+//international inch
+//https://en.wikipedia.org/wiki/Inch#Equivalences
+struct inch : named_scaled_unit<inch, "in", no_prefix, ratio<254,100,-2>, si::metre> {};
+
+//intrnational mile
+//https://en.wikipedia.org/wiki/Mile#International_mile
+struct mile : named_scaled_unit<mile, "mi", no_prefix, ratio<25'146,15'625,3>, si::metre> {};
+
+//international nautical mile
+//https://en.wikipedia.org/wiki/Mile#Nautical_mile
+struct nautical_mile : named_scaled_unit<nautical_mile,"mi(naut)",no_prefix, ratio<463,250,3>,si::metre> {};
+
+#if 0
+//https://en.wikipedia.org/wiki/Thousandth_of_an_inch
```

Comment:
Though they are the sam conversion factor, it should be possible for them to coexist, but get a compile error at the moment if I uncommet thou unit

Reply:
Yeah, downcasting facility again :-( If the framework will get `scaled_unit<ratio<127, 50, -5>, metre>` it will need only one type associated with it for downcasting. If we need different names we can just use an alias but the framework will always downcast to the primary one.

The problem starts to bother me because kelvin and Celsius have the same ratio too :-(

Any ideas?

Reply:
> Yeah, downcasting facility again :-( If the framework will get `scaled_unit<ratio<127, 50, -5>, metre>` it will need only one type associated with it for downcasting. If we need different names we can just use an alias but the framework will always downcast to the primary one.
> 
> The problem starts to bother me because kelvin and Celsius have the same ratio too :-(

How about a way to turn downcasting on and off? My choice would be default to off and then turn it on for quantities where it makes sense.

Reply:
Well, it makes sense for all derived units and this is how it works now. The only problem is when two equivalent units (the same ratio of a reference unit) have different names. In such a case treating them as an alias probably is not too far from reality. The framework will never have the ability to determine which one to use as a target so we have to choose one.

On the other hand, I will try to develop a mechanism that will disable downcasting if more than one conflicting declaration is defined. This means that the behavior will depend on the number of include files or imported modules. But for now, I do not if it is doable at, all...

Reply:
Though not quite this simple, surely it is doable...
```c++
#include <type_traits>

struct A{};
struct B{};

template<typename T>
struct want_downcast : std::true_type{};

template<typename T>
struct downcast{ };

template<typename T>
struct identity{ typedef T type;};

template <> struct want_downcast<B> : std::false_type{};

template<typename Child>
struct derived_dimension : std::conditional_t< 
   want_downcast<Child>::value,
   identity<downcast<Child> >,
   identity<Child>
>{};

int main()
{
   static_assert(std::is_same<
       typename derived_dimension<A>::type, downcast<A> >::value 
   ,"");
   static_assert(std::is_same<
       typename derived_dimension<B>::type, B >::value
    ,"");
}
```